fontra2fontir/src/source.rs

```diff
@@ -1,22 +1,146 @@
+use std::{
+    collections::HashMap,
+    fs::File,
+    io::{BufRead, BufReader},
+    path::PathBuf,
+};
 
-use fontir::source::Source;
+use fontdrasil::types::GlyphName;
+use fontir::{
+    error::Error,
+    source::{Input, Source},
+    stateset::StateSet,
+};
+use percent_encoding::{utf8_percent_encode, NON_ALPHANUMERIC};
 
 pub struct FontraIrSource {
-    _fontra_dir: PathBuf,
+    fontdata_file: PathBuf,
+    glyphinfo_file: PathBuf,
+    glyph_dir: PathBuf,
+    glyph_info: HashMap<GlyphName, (PathBuf, Option<u32>)>,
 }
 
 impl FontraIrSource {
-    pub fn new(fontra_dir: PathBuf) -> Self {
-        FontraIrSource {
-            _fontra_dir: fontra_dir,
+    pub fn new(fontra_dir: PathBuf) -> Result<Self, Error> {
+        let fontdata_file = fontra_dir.join("font-data.json");
+        if !fontdata_file.is_file() {
+            return Err(Error::FileExpected(fontdata_file));
         }
+        let glyphinfo_file = fontra_dir.join("glyph-info.csv");
+        if !glyphinfo_file.is_file() {
+            return Err(Error::FileExpected(glyphinfo_file));
+        }
+        let glyph_dir = fontra_dir.join("glyphs");
+        if !glyph_dir.is_dir() {
+            return Err(Error::DirectoryExpected(glyph_dir));
+        }
+        Ok(FontraIrSource {
+            fontdata_file,
+            glyphinfo_file,
+            glyph_dir,
+            glyph_info: Default::default(),
+        })
+    }
+
+    fn glyph_file(&self, glyph: GlyphName) -> PathBuf {
+        // TODO: this probably over-encodes. Should be sufficient to get started.
+        let filename = utf8_percent_encode(glyph.as_str(), NON_ALPHANUMERIC);
+        self.glyph_dir.join(filename.to_string() + ".json")
+    }
+
+    // When things like upem may have changed forget incremental and rebuild the whole thing
+    fn static_metadata_state(&self) -> Result<StateSet, Error> {
+        let mut font_info = StateSet::new();
+        font_info.track_file(&self.fontdata_file)?;
+        font_info.track_file(&self.glyphinfo_file)?;
+        Ok(font_info)
+    }
+
+    fn load_glyphinfo(&mut self) -> Result<(), Error> {
+        // Read the glyph-info file
+        let file = File::open(&self.glyphinfo_file).map_err(Error::IoError)?;
+
+        // Example files suggest the first line is just the column headers. Hopefully always :)
+        // This file is tool generated so it shouldn't be full of human error. Fail if we don't understand.
+        for (i, line) in BufReader::new(file).lines().enumerate().skip(1) {
+            let line = line.map_err(Error::IoError)?;
+            let parts: Vec<_> = line.split(';').collect();
+            if parts.len() != 2 {
+                return Err(Error::ParseError(
+                    self.glyphinfo_file.clone(),
+                    format!("Expected two parts in line {i} separated by ;"),
+                ));
+            }
+            let glyph_name = GlyphName::new(parts[0].trim());
+            // TODO: support multiple codepoints
```

Comment:
the mutliple codepoints are simply comma separated, see https://github.com/googlefonts/fontra/blob/a4edd06837118e583804fd963c22ed806a315b04/src/fontra/backends/fontra.py#L171-L181